File: phalcon/library/FileEo.php
<?php
namespace Library;

use Service\Base;

/* 文件类 */
class FileEo extends Base {

  public static $Root = '';

  /* 列表 */
  static function List(string $path=''): array {
    // 路径
    $path = $path=='/'?'':trim($path, '/').'/';
    $path = preg_replace('/\.\.|\.\\//','',$path);
    // 数据
    $res = ['path'=> $path, 'dirNum'=> 0, 'fileNum'=> 0, 'size'=> 0, 'folder'=> [], 'files'=> []];
    // 是否文件夹
    $root = self::$Root.$path;
    if(!is_dir($root)) return $res;
    // 文件夹&文件
    $list = scandir($root);
    foreach($list as $f) {
      if($f=='.' || $f=='..') continue;
      $ff = $root . '/' . $f;
      $size = self::FileSize($ff);
      $res['size'] += $size;
      $ctime = self::GetCtime($ff);
      $mtime = self::GetMtime($ff);
      $perm = self::GetPerm($ff);
      if(is_dir($ff)){
        $res['folder'][] = ['name'=>$f, 'size'=>self::FormatBytes($size), 'ctime'=>$ctime, 'mtime'=>$mtime, 'perm'=>$perm];
        $res['dirNum']++;
      }else {
        $ext = self::GetExt($f);
        $res['files'][] = ['name'=>$f, 'size'=>self::FormatBytes($size), 'ctime'=>$ctime, 'mtime'=>$mtime, 'perm'=>$perm, 'ext'=>$ext];
        $res['fileNum']++;
      }
    }
    // 大小
    $res['size'] = self::FormatBytes($res['size']);
    return $res;
  }

  /* 全部文件 */
  static function AllFile(string $path=''): array {
    $res = [];
    // 是否文件夹
    $root = self::$Root.$path;
    if(!is_dir($root)) return $res;
    // 文件夹&文件
    $list = scandir($root);
    foreach($list as $f) {
      if($f=='.' || $f=='..') continue;
      $ff = $root . '/' . $f;
      if(is_file($ff)) $res[] = $f;
    }
    return $res;
  }

  /* 统计大小 */
  static function FileSize($ff){
    $total = 0;
    // 文件
    if(is_file($ff)){
      $total += filesize($ff);
    }elseif(is_dir($ff)){
      // 文件夹
      $list = scandir($ff);
      foreach($list as $f){
        if($f=='.' || $f=='..') continue;
        $total += self::FileSize($ff.'/'.$f);
      }
    }
    return $total;
  }
  /* 创建时间 */
  static function GetCtime($ff) {
    return date("Y-m-d H:i:s",filectime($ff));
  }
  /* 修改时间 */
  static function GetMtime($ff='') {
    return date("Y-m-d H:i:s",filemtime($ff));
  }
  /* 获取权限值 */
  static function GetPerm($ff) {
    return substr(sprintf('%o',fileperms($ff)),-3);
  }
  /* 文件后缀 */
  static function GetExt($fileName){
    return strtolower(substr(strrchr($fileName, '.'), 1));
  }
  /* 格式化 */
  static function FormatBytes($bytes){
    if($bytes >= 1073741824){
      $bytes = round($bytes*100/1073741824)/100 . ' GB';
    }elseif($bytes >= 1048576){
      $bytes = round($bytes*100/1048576)/100 . ' MB';
    }elseif($bytes >= 1024){
      $bytes = round($bytes*100/1024)/100 . ' KB';
    }else{
      $bytes = $bytes . ' B';
    }
    return $bytes;
  }

  /* 是否文件 */
  static function IsFile(string $file=''): bool {
    $file = self::$Root.$file;
    return is_file($file);
  }

  /* 创建目录 */
  static function Mkdir(string $path=''): bool {
    $path = self::$Root.$path;
    if(!file_exists($path)) return mkdir($path,0777,true);
    return true;
  }
  static function MkdirFile(string $file=''): bool {
    if(empty($file)) return false;
    $arr = explode('/', $file);
    $n = count($arr);
    if($n>=2){
      unset($arr[$n-1]);
      $dir = implode('/', $arr);
      if(!file_exists($dir)) return mkdir($dir, 0777, true);
    }
    return true;
  }

  /* 重命名 */
  static function Rename(string $rename, string $name): bool {
    $src = self::$Root.$rename;
    $dst = self::$Root.$name;
    return rename($src,$dst);
  }

  /* 上传 */
  static function Upload(string $tmp, string $to): bool {
    $to = self::$Root.$to;
    return move_uploaded_file($tmp, $to);
  }

  /* 写入 */
  static function Writer(string $file='', string $content): bool {
    $file = self::$Root.$file;
    self::MkdirFile($file);
    return file_put_contents($file, $content)?true:false;
  }

  /* 追加 */
  static function WriterEnd(string $file='', string $content): bool {
    $file = self::$Root.$file;
    self::MkdirFile($file);
    return file_put_contents($file, $content, FILE_APPEND)?true:false;
  }

  /* 读取 */
  static function Bytes(string $file) {
    $dst = self::$Root.$file;
    return file_get_contents($dst);
  }

  /* 删除(文件夹&文件) */
  static function RemoveAll(string $path='') {
    if($path=='') return false;
    $obj = self::$Root.$path;
    if(is_dir($obj)){
      // 文件夹
      $list = scandir($obj);
      foreach ($list as $f) {
        if($f=='.' || $f=='..') continue;
        $ff = $path.'/'.$f;
        if(is_dir(self::$Root.$ff)) self::RemoveAll($ff);
        else unlink(self::$Root.$ff);
      }
      // 空目录
      rmdir($obj);
    }elseif(is_file($obj)){
      // 文件
      unlink($obj);
    }
  }

}

File: gin/library/file_eo.go
package library

import (
	"fmt"
	"io/ioutil"
	"mime/multipart"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"webmis/util"

	"github.com/gin-gonic/gin"
)

/* 文件类 */
type FileEo struct {
	Root string
}

/* 创建 */
func (fe *FileEo) New(root string) *FileEo {
	fe.Root = root
	return fe
}

/* 列表 */
func (fe FileEo) List(path string) map[string]interface{} {
	// 路径
	if path == "/" {
		path = ""
	} else {
		path = util.Trim(path, "/") + "/"
	}
	reg := regexp.MustCompile("\\.\\.|\\.\\/")
	path = reg.ReplaceAllString(path, "")
	// 数据
	folder := []map[string]interface{}{}
	files := []map[string]interface{}{}
	res := map[string]interface{}{
		"path":    path,
		"dirNum":  0,
		"fileNum": 0,
		"size":    0,
		"folder":  folder,
		"files":   files,
	}
	// 是否文件夹
	root := fe.Root + path
	dir, err := os.Stat(root)
	if err != nil {
		return res
	}
	if !dir.IsDir() {
		return res
	}
	// 文件夹&文件
	var total int64
	list, _ := ioutil.ReadDir(root)
	for _, f := range list {
		ff := root + "/" + f.Name()
		size := fe.FileSize(ff)
		total += size
		ctime := fe.GetCtime(ff)
		mtime := fe.GetMtime(ff)
		perm := fe.GetPerm(ff)
		if f.IsDir() {
			folder = append(folder, map[string]interface{}{
				"name":  f.Name(),
				"size":  fe.FormatBytes(size),
				"ctime": ctime,
				"mtime": mtime,
				"perm":  perm,
			})
		} else {
			files = append(files, map[string]interface{}{
				"name":  f.Name(),
				"size":  fe.FormatBytes(size),
				"ctime": ctime,
				"mtime": mtime,
				"perm":  perm,
				"ext":   fe.GetExt(f.Name()),
			})
		}
	}
	res["folder"] = folder
	res["files"] = files
	res["size"] = fe.FormatBytes(total)
	return res
}

/* 全部文件 */
func (fe FileEo) AllFile(path string) []string {
	res := []string{}
	// 是否文件夹
	root := fe.Root + path
	dir, err := os.Stat(root)
	if err != nil {
		return res
	}
	if !dir.IsDir() {
		return res
	}
	// 文件夹&文件
	list, _ := ioutil.ReadDir(root)
	for _, f := range list {
		if !f.IsDir() {
			res = append(res, f.Name())
		}
	}
	return res
}

/* 统计大小 */
func (fe FileEo) FileSize(ff string) int64 {
	var total int64
	// 文件
	dir, _ := os.Stat(ff)
	if !dir.IsDir() {
		return dir.Size()
	}
	// 文件夹
	list, _ := ioutil.ReadDir(ff)
	for _, f := range list {
		if f.IsDir() {
			dir := filepath.Join(ff, f.Name())
			total += fe.FileSize(dir)
		} else {
			total += f.Size()
		}
	}
	return total
}

/* 创建时间 */
func (FileEo) GetCtime(ff string) string {
	f, _ := os.Stat(ff)
	return f.ModTime().Format("2006-01-02 15:04:05")
}

/* 修改时间 */
func (FileEo) GetMtime(ff string) string {
	f, _ := os.Stat(ff)
	return f.ModTime().Format("2006-01-02 15:04:05")
}

/* 权限值 */
func (fe FileEo) GetPerm(ff string) int {
	f, _ := os.Stat(ff)
	perm := f.Mode().String()
	p1 := fe.permToVal(perm[1:4])
	p2 := fe.permToVal(perm[4:7])
	p3 := fe.permToVal(perm[7:10])
	res, _ := strconv.Atoi(p1 + p2 + p3)
	return res
}
func (FileEo) permToVal(perm string) string {
	var num int
	if perm[0:1] == "r" {
		num += 4
	}
	if perm[1:2] == "w" {
		num += 2
	}
	if perm[2:3] == "x" {
		num += 1
	}
	return strconv.Itoa(num)
}

/* 文件后缀 */
func (FileEo) GetExt(fileName string) string {
	arr := strings.Split(fileName, ".")
	if len(arr) > 0 {
		return arr[len(arr)-1:][0]
	}
	return ""
}

/* 格式化 */
func (FileEo) FormatBytes(bytes int64) string {
	var str string
	if bytes >= 1073741824 {
		str = fmt.Sprintf("%.2f GB", float64(bytes)/1073741824)
	} else if bytes >= 1048576 {
		str = fmt.Sprintf("%.2f MB", float64(bytes)/1048576)
	} else if bytes >= 1024 {
		str = fmt.Sprintf("%.2f KB", float64(bytes)/1024)
	} else {
		str = fmt.Sprintf("%.0f B", float64(bytes))
	}
	return str
}

/* 是否文件 */
func (fe FileEo) IsFile(file string) bool {
	file = fe.Root + file
	_, err := os.Stat(file)
	if err != nil {
		return false
	}
	return true
}

/* 创建目录 */
func (fe FileEo) Mkdir(path string) bool {
	path = fe.Root + path
	if err := os.MkdirAll(path, 0766); err != nil {
		return false
	}
	return true
}

/* 重命名 */
func (fe FileEo) Rename(rename string, name string) bool {
	src := fe.Root + rename
	dst := fe.Root + name
	if err := os.Rename(src, dst); err != nil {
		return false
	}
	return true
}

/* 上传 */
func (fe FileEo) Upload(c *gin.Context, file *multipart.FileHeader, filename string) bool {
	dst := fe.Root + filename
	if err := c.SaveUploadedFile(file, dst); err != nil {
		return false
	}
	return true
}

/* 写入 */
func (fe FileEo) Writer(file string, content string) error {
	file = fe.Root + file
	f, err := os.OpenFile(file, os.O_CREATE|os.O_RDWR, 0766)
	if err != nil {
		return err
	}
	defer f.Close()
	f.WriteString(content)
	return nil
}

/* 追加 */
func (fe FileEo) WriterEnd(file string, content string) error {
	file = fe.Root + file
	f, err := os.OpenFile(file, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0766)
	if err != nil {
		return err
	}
	defer f.Close()
	f.WriteString(content)
	return nil
}

/* 读取 */
func (fe FileEo) Bytes(file string) []byte {
	dst := fe.Root + file
	bytes, _ := ioutil.ReadFile(dst)
	return bytes
}

/* 删除(文件夹&文件) */
func (fe FileEo) RemoveAll(path string) error {
	path = fe.Root + path
	return os.RemoveAll(path)
}
